File: namastack-outbox-examples/namastack-outbox-example-java/src/main/java/io/namastack/demo/CustomerRegisteredOutboxHandler.java
package io.namastack.demo;

import io.namastack.demo.customer.CustomerRegisteredEvent;
import io.namastack.outbox.handler.OutboxHandlerIdentity;
import io.namastack.outbox.handler.OutboxRecordMetadata;
import io.namastack.outbox.handler.OutboxTypedHandler;
import org.jspecify.annotations.NonNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

@Component
public class CustomerRegisteredOutboxHandler implements OutboxTypedHandler<CustomerRegisteredEvent> {

  private static final Logger logger = LoggerFactory.getLogger(CustomerRegisteredOutboxHandler.class);

  @Override
  public OutboxHandlerIdentity getTypedHandlerIdentity() {
    return new OutboxHandlerIdentity("customers.send-registration-email");
  }

  @Override
  public void handle(CustomerRegisteredEvent payload, @NonNull OutboxRecordMetadata metadata) {
    logger.info("[Handler] Send email to: {}", payload.getEmail());
    ExternalMailService.send(payload.getEmail());
  }
}

File: namastack-outbox-examples/namastack-outbox-example-flyway-jdbc/src/main/kotlin/io/namastack/demo/CustomerRegisteredOutboxHandler.kt
package io.namastack.demo

import io.namastack.demo.customer.CustomerRegisteredEvent
import io.namastack.outbox.handler.OutboxRecordMetadata
import io.namastack.outbox.handler.OutboxTypedHandler
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class CustomerRegisteredOutboxHandler : OutboxTypedHandler<CustomerRegisteredEvent> {
    private val logger = LoggerFactory.getLogger(CustomerRegisteredOutboxHandler::class.java)

    override fun handle(
        payload: CustomerRegisteredEvent,
        metadata: OutboxRecordMetadata,
    ) {
        logger.info("[Handler] Send email to: {}", payload.email)
        ExternalMailService.send(payload.email)
    }
}
